12 finish request

whyamiroot opened the pull request:
SessionController - FinishSession checks session ID
obtained from JSON and returns 400 if null; checks equality of obtained
session ID and the one from DB and returns 400 if they are not equal;
checks session status and returns 409 if session if already finished.
Updates session status to FINSIHED and returns 200 OK.

Commit:
Add SessionController and FinishRequest.

Add stub method FinishSession, which returns 200 OK

whyamiroot committed:
Updated FinishRequest, FinishSession

Removed "finished" field from expected JSON, sessionId changed to
numeric. Started to implement FinishSession(waiting APIs from DB team)

whyamiroot committed:
Merge branch 'master' into 12-finish-request

whyamiroot committed:
Resolves #12. FinishSession finishes session

Updated SessionController - FinishSession checks session ID
obtained from JSON and returns 400 if null; checks equality of obtained
session ID and the one from DB and returns 409 if they are not equal;
checks session status and returns 409 if session if already finished.
Updates session status to FINSIHED and returns 200 OK.

whyamiroot committed:
Merge branch 'master' into 12-finish-request

whyamiroot commented on Comptech.Backend/src/Comptech.Backend.Service/Controllers/SessionController.cs:
```diff
@@ -0,0 +1,71 @@
+        private readonly IRepository<Session> repository;
+
+        public SessionController(UserManager<ApplicationUser> userManager,
+                               ILoggerFactory loggerFactory,
```
Возможно здесь нужно добавить у интерфейсов [FromServices]?

alex-ks replied:
Нет, в конструкторе это не нужно.

alex-ks commented on Comptech.Backend/src/Comptech.Backend.Service/Controllers/SessionController.cs:
```diff
@@ -0,0 +1,71 @@
+                if (request.SessionId == null)
+                {
+                    logger.LogError($"Attempted to finish session with null sessionId");
+                    return BadRequest();
```
Предлагаю вернуть какое-либо текстовое описание ошибки

alex-ks replied:
То же касается других вызовов BadRequest ниже.

alex-ks commented on Comptech.Backend/src/Comptech.Backend.Service/Controllers/SessionController.cs:
```diff
@@ -0,0 +1,71 @@
+
+        [Route("/rest/finished")]
+        [HttpPost]
+        public async Task<IActionResult> FinishSession([FromBody] FinishRequest request)
```
Не хватает обработки исключений. Думаю, здесь достаточно общего try-catch с логгированием Message и StackTrace, и возвратом BadRequest.

whyamiroot replied:
Всю функцию завернуть в try/catch?

alex-ks replied:
Да, этого хватит.

alex-ks reviewed (commented):
Не хватает сообщений об ошибках и обработки исключений.

whyamiroot committed:
Add Exception, BadRequests message

After-review changes. Add Exception handling - log message and stack
trace and send back BadRequest with exception message. Add error
messages to BadRequest response body.

whyamiroot committed:
Changed logger scope from UserController to SessionController

DVEfremov reviewed (commented):
looks good for me 
please answer questions

sumboid commented on Comptech.Backend/src/Comptech.Backend.Service/Controllers/SessionController.cs:
```diff
@@ -0,0 +1,92 @@
+            {
+                using (logger.BeginScope(nameof(FinishSession)))
+                {
+                    //если в запросе нет id сессии, то возвращаем ошибку(?)
```
Здесь и ниже есть комментарии, которые:
1. Довольно избыточные с точки зрения описания. По коду итак можно сходу понять, что происходит.
2. На русском языке, что значительно сужает круг лиц, которые смогут без переводчика их понять :)

Ну и плюс ко всему, комментарии -- это элемент, который приходится поддерживать вместе с кодом. Поэтому лучше их использовать в местах, где действительно необходимо пояснение.
